fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/UnnestedRecordTypeBuilder.java

```diff
@@ -0,0 +1,290 @@
+
+        List<Nesting> foundNestings = new ArrayList<>(typeProto.getNestingsCount());
+        for (RecordMetaDataProto.UnnestedRecordType.Nesting nestingProto : typeProto.getNestingsList()) {
+            String parent = nestingProto.getParent().isEmpty() ? UnnestedRecordType.PARENT_CONSTITUENT : nestingProto.getParent();
+            String child = nestingProto.getChild().isEmpty() ? UnnestedRecordType.PARENT_CONSTITUENT : nestingProto.getChild();
+            foundNestings.add(new Nesting(parent, child, KeyExpression.fromProto(nestingProto.getNestingExpression())));
+        }
+        this.nestings = foundNestings;
+    }
+
+    @Nullable
+    private static Descriptors.Descriptor findDescriptorByName(@Nonnull Descriptors.FileDescriptor fileDescriptor, @Nonnull String fullName) {
+        // Use the seen set to protect against circular dependencies. Files should be added to the set right before they are searched through, so
+        // each file will be visited at most once
+        Set<Descriptors.FileDescriptor> seen = new HashSet<>();
+        seen.add(fileDescriptor);
+        return findDescriptorByName(fileDescriptor, fullName, seen);
+    }
+
+    @Nullable
+    private static Descriptors.Descriptor findDescriptorByName(@Nonnull Descriptors.FileDescriptor fileDescriptor, @Nonnull String fullName, @Nonnull Set<Descriptors.FileDescriptor> seen) {
+        if (fullName.startsWith(fileDescriptor.getPackage())) {
+            for (Descriptors.Descriptor desc : fileDescriptor.getMessageTypes()) {
+                Descriptors.Descriptor found = findDescriptorByName(desc, fullName);
+                if (found != null) {
+                    return found;
+                }
+            }
+        }
+        for (Descriptors.FileDescriptor dependency : fileDescriptor.getDependencies()) {
+            if (seen.add(dependency)) {
+                Descriptors.Descriptor found = findDescriptorByName(dependency, fullName, seen);
+                if (found != null) {
+                    return found;
+                }
+            }
+        }
+        for (Descriptors.FileDescriptor publicDependency : fileDescriptor.getPublicDependencies()) {
+            if (seen.add(publicDependency)) {
+                Descriptors.Descriptor found = findDescriptorByName(publicDependency, fullName, seen);
+                if (found != null) {
+                    return found;
+                }
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Descriptors.Descriptor findDescriptorByName(@Nonnull Descriptors.Descriptor descriptor, @Nonnull String fullName) {
+        if (fullName.startsWith(descriptor.getFullName())) {
+            if (descriptor.getFullName().equals(fullName)) {
+                return descriptor;
+            }
+            for (Descriptors.Descriptor nestedType : descriptor.getNestedTypes()) {
+                Descriptors.Descriptor foundNested = findDescriptorByName(nestedType, fullName);
```

Comment:
I found a problem with this piece of code:
message MapRecord {
  int64 id = 1;
  MapFromStringToString map_field = 2;
}
getNestedTypes() returns empty list. Maybe "nestedTypes" here only mean message types defined inside MapRecord.
We should probably do:
for (FieldDescriptor field: descriptor.getFields()) {
  if (field.getType() == MESSAGE) {
    Descriptor foundNested = findDescriptorByName(field.getDescriptor(), fullName);
    if (foundNested != null) { return foundNested; }
  }
}

Reply:
Yeah, `descriptor.getNestedTypes()` returns the list of nested messages defined within the type. The goal of this code is to find, by iterating through the list of defined message types, is to find the type that matches the given name. So, if you had something like:

```proto
package com.example.package;

message MapFromStringToString {
    message Entry {
        optional string key = 1;
        optional string value = 2;
    }
    repeated Entry entry = 1;
}

message MapRecord {
    int64 id = 1;
    MapFromStringToString map_field = 2;
}
```

Then I'd expect the `fullName` to be something like `com.example.package.MapFromStringToString.Entry`. When it gets to the `MapRecord`, that name shouldn't match, and it should move on. When it gets to `MapFromStringToString`, it should look at its nested types to find the correct `Entry` type.

Importantly, this can be defined within the same stored record type or outside of it, and it can be within the same file (as in this example) or in one of the dependencies. If you have a potential example of full name and file descriptor that it fails to find, though, that would be good to know.